apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RetryOptions.cs

```diff
@@ -0,0 +1,127 @@
+    public enum ExceptionHandling
+    {
+        /// <summary>
+        /// Let the errors propagate up after they have been retried if they continually fail.
+        /// This may kill the running process.
+        /// </summary>
+        Propagate,
+
+        /// <summary>Swallow and ignore the errors.</summary>
+        Swallow,
+    }
+
+    /// <summary>
+    /// Options that define retry logic for requests that failed.
+    /// </summary>
+    public sealed class RetryOptions
+    {
+        /// <summary>The default buffer size in bytes. 2^18 = 262144.</summary>
+        private const int DefaultBufferSize = 262144;
+
+        /// <summary>The default number of retry attempts.</summary>
+        private const int DefaultRetryAttempts = 10;
+
+        /// <summary>The default wait time between retries.</summary>
+        private static readonly TimeSpan DefaultRetryInterval = TimeSpan.FromSeconds(2);
+
+        /// <summary>The type of retry to make.</summary>
+        public RetryType RetryType { get; }
+
+        /// <summary>How exceptions are handled.</summary>
+        public ExceptionHandling ExceptionHandling { get; }
+
+        /// <summary>How to handle a full buffer.</summary>
+        public BufferOverflow BufferOverflow { get;  }
+
+        /// <summary>The size of the buffer in bytes, only used for <see cref="RetryType.Retry"/></summary>
+        public int BufferSizeBytes { get; }
+
+        /// <summary>The number of attempts to retry sending information, only used for <see cref="RetryType.Retry"/></summary>
+        public int RetryAttempts { get; }
+
+        /// <summary>The amount of time to wait between retries, only used for <see cref="RetryType.Retry"/></summary>
+        public TimeSpan RetryInterval { get; }
+
+        internal RetryOptions(RetryType retryType, ExceptionHandling exceptionHandling, BufferOverflow bufferOverflow,
+            int? bufferSizeBytes = null, int? retryAttempts = null, TimeSpan? retryInterval = null)
+        {
+            RetryType = GaxPreconditions.CheckEnumValue(retryType, nameof(retryType));
+            ExceptionHandling = GaxPreconditions.CheckEnumValue(exceptionHandling, nameof(exceptionHandling));
+            BufferOverflow = GaxPreconditions.CheckEnumValue(bufferOverflow, nameof(bufferOverflow));
+            BufferSizeBytes = GaxPreconditions.CheckArgumentRange(
+                bufferSizeBytes ?? default(int), nameof(bufferSizeBytes), 0, int.MaxValue);
+            RetryAttempts = GaxPreconditions.CheckArgumentRange(
+                retryAttempts ?? default(int), nameof(retryAttempts), 0, int.MaxValue);
+            RetryInterval = retryInterval ?? default(TimeSpan);
+            GaxPreconditions.CheckArgument(RetryInterval >= TimeSpan.Zero, nameof(retryInterval), 
+                $"{nameof(retryInterval)} must be greater than 0");
+        }
+
+        /// <summary>
+        /// Create a <see cref="RetryOptions"/> for <see cref="RetryType.None"/>.
```

Comment:
I would use the descriptive "Creates" here (and below) rather than the imperative "Create" to be consistent with other summaries.